File: leetcode/problem_238.py
# Problem 238 (Medium): https://leetcode.com/problems/product-of-array-except-self/

# Solution 1 uses division operator

class Solution:
    def productExceptSelf(self, nums):
        product = 1
        result = []
        nonZeroProduct = 0
        zerocount = 0
        for i in range(len(nums)):
            if(nums[i] != 0):
                if(nonZeroProduct == 0):
                    nonZeroProduct = 1
                nonZeroProduct *=nums[i]
            else:
                zerocount+=1
            product *= nums[i]

        for i in range(len(nums)):
            if(nums[i] == 0):
                if(zerocount > 1):
                    result.append(product)
                else:
                    result.append(nonZeroProduct)
            else:
                result.append(product//nums[i])
        return result

class Solution2:
    def productExceptSelf(self, nums):
        pref = [1 for i in range(len(nums))]
        suf = [1 for i in range(len(nums))]
        result = []
        for i in range(1, len(nums)):
            pref[i] = (pref[i-1] * nums[i-1])

        for i in range(len(nums)-2, -1, -1):
            suf[i] = (suf[i+1] * nums[i+1])
        for i in range(len(nums)):
            result.append(pref[i]*suf[i])
        return result


if __name__ == "__main__":
    nums = [1,2,3,4]
    print(Solution().productExceptSelf(nums)) # [24,12,8,6]

File: leetcode/problem_238.kt
/**
 * Problem 238. Product of Array Except Self
 * tags: array, prefix sum
 */
// time complexity: Bruteforce O(n^2)
// space complexity: O(1)
class Solution {
    fun productExceptSelf(nums: IntArray): IntArray {
        /*
        *  input: [1, 2, 3, 4, 5]
         * example: [1, 2, 6, 24, 120]
         */
        var result: IntArray = IntArray(nums.size){0}
        for( i in nums.indices){
            var product = 1
            for(j in nums.indices){
                if(i != j){
                    product *= nums[j]
                }
            }
            result[i] = product
        }
        return result  
    }
}


// time complexity: O(n)
// space complexity: O(n)
class OptimizedSolution {
    fun productExceptSelf(nums: IntArray): IntArray {
        /*
        *  input: [1, 2, 3, 4, 5]
         * example: [1, 2, 6, 24, 120]
         * 
         */
        var result: IntArray = IntArray(nums.size){0}
        var prefix: IntArray = IntArray(nums.size){0}
        var postfix: IntArray = IntArray(nums.size){0}
        var product = 1
        var postProduct = 1
        for(i in nums.indices){
            product *= nums[i]
            prefix[i] = product
            var j = (nums.size - 1) - i
            postProduct *= nums[j]
            postfix[j] = postProduct
        }
        for(i in nums.indices){
            if(i == 0){
                result[i] = postfix[i + 1]
            } else if(i == nums.size - 1){
                result[i] = prefix[i - 1]
            } else{
                result[i] = prefix[i-1] * postfix[i + 1]
            }
        }
        return result  
    }
}

// time complexity: O(n)
// space complexity: O(1)
class Solution {
    fun productExceptSelf(nums: IntArray): IntArray {
        /*
        *  input: [1, 2, 3, 4, 5]
         * example: [1, 2, 6, 24, 120]
         * 
         */
        var result: IntArray = IntArray(nums.size){0}
        var product = 1
        // store prefix in the array
        for(i in nums.indices){
            result[i] = product
            product *= nums[i]
        }
        product = 1
        for(i in nums.size - 1 downTo 0){
            result[i] *= product
            product *= nums[i]
        }
        return result  
    }
}